geomstats/learning/frechet_mean.py

```diff
@@ -72,59 +67,55 @@ def linear_mean(points, weights=None):
     mean : array-like, shape=[1, dimension]
         Weighted linear mean of the points.
     """
+    # TODO(ninamiolane): Factorize this code to handle lists
+    # in the whole codebase
     if isinstance(points, list):
-        points = gs.vstack(points)
-    points = gs.to_ndarray(points, to_ndim=2)
-    n_points = gs.shape(points)[0]
-
+        points = gs.stack(points, axis=0)
     if isinstance(weights, list):
-        weights = gs.vstack(weights)
-    elif weights is None:
-        weights = gs.ones((n_points, ))
+        weights = gs.stack(weights, axis=0)
+
+    n_points = geomstats.vectorization.get_n_points(
+        points, point_type)
+
+    if weights is None:
+        weights = gs.ones((n_points,))
 
     weighted_points = gs.einsum('...,...j->...j', weights, points)
-    mean = (gs.sum(weighted_points, axis=0) / gs.sum(weights))
-    mean = gs.to_ndarray(mean, to_ndim=2)
+    mean = gs.sum(weighted_points, axis=0) / gs.sum(weights)
     return mean
 
 
 def _default_gradient_descent(points, metric, weights,
                               max_iter, point_type, epsilon, verbose):
     if point_type == 'vector':
         points = gs.to_ndarray(points, to_ndim=2)
-        einsum_str = 'nk,nj->j'
+        einsum_str = 'n,nj->j'
```

Comment:
why not '...' instead of to_ndarray?

Reply:
In this case, using to_ndarray allows extracting the n_points (`gs.shape(points)[0]`) and the init mean (`points[0]`) directly and no end-of-function reshaping is needed. Otherwise, some if/else checks have to be done, for example, like in `get_n_points`, and I think this would be actually slower.